### PR TITLE
PRIME-2769 Archive Site

### PR DESCRIPTION
**Pull request by bergomi02**

### CODE REVIEW

File: prime-dotnet-webapi-tests/Unit Tests/SiteStatusStateEngine.Tests.cs [hunk 1]
```diff
@@ -21,13 +21,31 @@ public static IEnumerable<object[]> SiteRegistrationActionData()
 
         [Theory]
         [MemberData(nameof(SiteRegistrationActionData))]
-        public void TestAllowableStatusChange_Active(SiteRegistrationAction action)
+        public void TestAllowableStatusChange_Actived(SiteRegistrationAction action)
```
Comment (neophyte57) [inline prime-dotnet-webapi-tests/Unit Tests/SiteStatusStateEngine.Tests.cs:24]: Should be `TestAllowableStatusChange_Archived`

File: prime-dotnet-webapi-tests/Unit Tests/SiteStatusStateEngine.Tests.cs [hunk 2]
```diff
@@ -21,13 +21,31 @@ public static IEnumerable<object[]> SiteRegistrationActionData()
...
+            // Act
+            var allowed = SiteStatusStateEngine.AllowableStatusChange(action, SiteStatusType.Archived);
+
+            // Assert
+            Assert.Equal(expectedAllowed, allowed);
+        }
+
+        [Theory]
+        [MemberData(nameof(SiteRegistrationActionData))]
+        public void TestAllowableStatusChange_Archive(SiteRegistrationAction action)
```
Comment (neophyte57) [inline prime-dotnet-webapi-tests/Unit Tests/SiteStatusStateEngine.Tests.cs:41]: To fit the existing convention, should be `TestAllowableStatusChange_Editable`

File: prime-dotnet-webapi/Controllers/SitesController.cs [hunk 3]
```diff
@@ -14,6 +14,7 @@
 using Prime.Services;
 using Prime.ViewModels;
 using Prime.ViewModels.Sites;
+using RazorEngine.Compilation.ImpromptuInterface.InvokeExt;
```
Comment (neophyte57) [inline prime-dotnet-webapi/Controllers/SitesController.cs:17]: Extraneous

Comment (bergomi02) [reply prime-dotnet-webapi/Controllers/SitesController.cs]: How it got there? it is removed.

File: prime-dotnet-webapi-tests/Unit Tests/SiteStatusStateEngine.Tests.cs (hunk 1 above)
Comment (bergomi02) [reply prime-dotnet-webapi-tests/Unit Tests/SiteStatusStateEngine.Tests.cs]: Correct. I just renamed the test methods to make more sense.

File: prime-dotnet-webapi-tests/Unit Tests/SiteStatusStateEngine.Tests.cs (hunk 2 above)
Comment (bergomi02) [reply prime-dotnet-webapi-tests/Unit Tests/SiteStatusStateEngine.Tests.cs]: Correct. I just renamed the test methods to make more sense.

**Comment by neophyte57** [comment]
It seems if I try to Restore "Site 3A", it should work rather than getting an error message?

image

File: prime-dotnet-webapi/Controllers/SitesController.cs [hunk 4]
```diff
@@ -1382,5 +1382,88 @@ public async Task<ActionResult> GetDeviceProviderSite(string deviceProviderId)
...
+
+
+        // POST: api/Sites/5/archive
+        /// <summary>
+        /// Archive a site
+        /// </summary>
+        /// <param name="siteId"></param>
+        /// <param name="siteNoteUpdateModel"></param>
+        [HttpPost("{siteId}/archive", Name = nameof(ArchiveSite))]
+        [Authorize(Roles = Roles.EditSite)]
```
Comment (neophyte57) [inline prime-dotnet-webapi/Controllers/SitesController.cs:1394]: These should match what is in `site-registration-actions.component.html`

Comment (bergomi02) [reply prime-dotnet-webapi/Controllers/SitesController.cs:1394]: Sorry, what should match what?

Comment (neophyte57) [reply prime-dotnet-webapi/Controllers/SitesController.cs:1394]: Requiring Super Admin role

**Comment by bergomi02** [comment]
> It seems if I try to Restore "Site 3A", it should work rather than getting an error message?
> 
> image

It is fixed. Also, there was a bug where the logic is not putting "in-review status" as active site into account.

File: prime-dotnet-webapi/Services/interfaces/ISiteService.cs [hunk 5]
```diff
@@ -51,5 +51,9 @@ public interface ISiteService
         Task<SiteAdjudicationDocument> AddSiteAdjudicationDocumentAsync(int siteId, Guid documentGuid, int adminId);
         Task<IEnumerable<SiteAdjudicationDocument>> GetSiteAdjudicationDocumentsAsync(int siteId);
         Task<string> GetSitePecAsync(int siteId);
+        Task<List<Site>> GetSiteByPecAsync(string pec);
+        Task ArchiveSite(int siteId, string note);
+        Task RestoreArchivedSite(int siteId, string note);
+        Task<bool> CanBeRestore(int siteId);
```
Comment (neophyte57) [inline prime-dotnet-webapi/Services/interfaces/ISiteService.cs:57]: Should be `CanBeRestored`

Comment (bergomi02) [reply prime-dotnet-webapi/Services/interfaces/ISiteService.cs]: Updated

**Comment by neophyte57** [comment]
Should be hidden from Authorized User?

image

**Comment by bergomi02** [comment]
> Should be hidden from Authorized User?
> 
> image

it is fixed.

**Comment by neophyte57** [comment]
I tried to Restore a HA Site that had no Site ID but got error message 

image

**Comment by neophyte57** [comment]
> I tried to Restore a HA Site that had no Site ID but got error message
> 
> image

Can reproduce in PR environment

image

File: prime-angular-frontend/src/app/modules/health-auth-site-reg/pages/site-management-page/site-management-page.component.html [hunk 6]
```diff
@@ -110,6 +110,15 @@
...
       </ng-container>
+      <ng-container *ngIf="healthAuthoritySite.isArchived()"
+                    [ngTemplateOutlet]="notification"
+                    [ngTemplateOutletContext]="{
+                      props: {
+                        icon: 'not_interested',
+                        text: 'Archived'
+                      }
+                    }">
+      </ng-container>
```
Comment (neophyte57) [inline prime-angular-frontend/src/app/modules/health-auth-site-reg/pages/site-management-page/site-management-page.component.html:121]: Is this code necessary?

File: prime-angular-frontend/src/app/modules/site-registration/pages/site-management-page/site-management-page.component.html [hunk 7]
```diff
@@ -130,6 +130,15 @@
...
         </ng-container>
+        <ng-container *ngIf="isArchived(site)"
+                      [ngTemplateOutlet]="notification"
+                      [ngTemplateOutletContext]="{
+                        props: {
+                        icon: 'not_interested',
+                        text: 'Archived'
+                        }
+                        }">
+        </ng-container>
```
Comment (neophyte57) [inline prime-angular-frontend/src/app/modules/site-registration/pages/site-management-page/site-management-page.component.html:141]: Necessary?

File: prime-angular-frontend/src/app/modules/site-registration/pages/site-management-page/site-management-page.component.html [hunk 8]
```diff
@@ -221,7 +230,8 @@
       </button>
       <button *ngIf="site.careSettingCode === CareSettingEnum.PRIVATE_COMMUNITY_HEALTH_PRACTICE"
               mat-menu-item
-              [disabled]="!site.completed || site.status === SiteStatusType.IN_REVIEW"
+              [disabled]="!site.completed || site.status === SiteStatusType.IN_REVIEW
+              || site.status === SiteStatusType.ARCHIVED"
```
Comment (neophyte57) [inline prime-angular-frontend/src/app/modules/site-registration/pages/site-management-page/site-management-page.component.html:234]: Necessary?

**Comment by bergomi02** [comment]
> > I tried to Restore a HA Site that had no Site ID but got error message
> > image
> 
> Can reproduce in PR environment
> 
> image

Good catch! it is fixed.

File: prime-angular-frontend/src/app/modules/health-auth-site-reg/pages/site-management-page/site-management-page.component.html (hunk 6 above)
Comment (bergomi02) [reply prime-angular-frontend/src/app/modules/health-auth-site-reg/pages/site-management-page/site-management-page.component.html]: Other good catch. Removed.

File: prime-angular-frontend/src/app/modules/site-registration/pages/site-management-page/site-management-page.component.html (hunk 7 above)
Comment (bergomi02) [reply prime-angular-frontend/src/app/modules/site-registration/pages/site-management-page/site-management-page.component.html]: Other good catch. Removed

File: prime-angular-frontend/src/app/modules/site-registration/pages/site-management-page/site-management-page.component.html (hunk 8 above)
Comment (bergomi02) [reply prime-angular-frontend/src/app/modules/site-registration/pages/site-management-page/site-management-page.component.html]: No, removed.

File: prime-angular-frontend/src/app/shared/components/dialogs/content/site-archive-restore/site-archive-restore.component.ts [hunk 9]
```diff
@@ -0,0 +1,87 @@
+import { Component, OnInit, Inject, Output, EventEmitter } from '@angular/core';
```
[the remark this answers is not in the document]
Comment (bergomi02) [reply prime-angular-frontend/src/app/shared/components/dialogs/content/site-archive-restore/site-archive-restore.component.ts]: removed

File: prime-dotnet-webapi/Services/SiteService.cs [hunk 10]
```diff
@@ -14,6 +14,7 @@
 using System.Security.Claims;
 using Prime.HttpClients;
 using DelegateDecompiler.EntityFrameworkCore;
+using RazorEngine.Compilation.ImpromptuInterface.InvokeExt;
```
Comment (neophyte57) [inline prime-dotnet-webapi/Services/SiteService.cs:17]: Unused

File: prime-dotnet-webapi/Services/interfaces/ISiteService.cs [hunk 11]
```diff
@@ -51,5 +51,9 @@ public interface ISiteService
         Task<SiteAdjudicationDocument> AddSiteAdjudicationDocumentAsync(int siteId, Guid documentGuid, int adminId);
         Task<IEnumerable<SiteAdjudicationDocument>> GetSiteAdjudicationDocumentsAsync(int siteId);
         Task<string> GetSitePecAsync(int siteId);
+        Task<List<Site>> GetSiteByPecAsync(string pec);
```
Comment (neophyte57) [inline prime-dotnet-webapi/Services/interfaces/ISiteService.cs:54]: Should be `GetSitesByPecAsync`

**Review by neophyte57 — APPROVED** [review]
Whew, good work!

File: prime-dotnet-webapi/Services/interfaces/ISiteService.cs (hunk 11 above)
Comment (bergomi02) [reply prime-dotnet-webapi/Services/interfaces/ISiteService.cs]: fixed

File: prime-dotnet-webapi/Services/SiteService.cs (hunk 10 above)
Comment (bergomi02) [reply prime-dotnet-webapi/Services/SiteService.cs]: removed

**Pull request merged**